### ISSUE TITLE
Traceback error - servolid.py

### ISSUE THREAD

**Issue by JashwanthKumar252** [JashwanthKumar252/Autonomous_robot#2]
While trying to run the servolid.py code to get the data from the lidar , it was throwing the following errors:

1. The Lidar doesn't post any data at all and only the flag variable "p" gets printed. 

![Error_1](https://user-images.githubusercontent.com/81487333/124246577-e6342b80-dad5-11eb-932f-3a3f97a5a03a.png)

2. The Lidar fails to post data at random steps of the servo . (Only the variable "p" is printed and no data is posted),

![Error_2](https://user-images.githubusercontent.com/81487333/124246620-f0eec080-dad5-11eb-9c98-d10432ea106d.png)

3. The Lidar posts data for sometime and the traceback error pops and the file stops running.

![Error_3](https://user-images.githubusercontent.com/81487333/124246641-f4824780-dad5-11eb-99ed-5773bf4ac542.png)

All these 3 errors, occur randomly and at no order during the execution of the code. To simulate the error run the following command from the respective directory of servolid.py
```
sudo python3 servolid.py

```
@pradhanshrijal  @Dcruise546

**Comment by pradhanshrijal** [comment]
Your serial connect is not stable.  Due to this there are certain instances where the data is not obtained.

A good approach would be to use error and exception handles whenever you communicate with the serial Port.

Can you try to set this up and test again?

**Comment by JashwanthKumar252** [comment]
As said we first used Try and Except inside the "Lidar"  Class in servolid.py where if we enter "except" we assigned  a value of 0. But only the value of "p" got printed and not even the value of zero that we assigned in the exception was printed. This means that the "lidar.lidar_read()" inside the while loop is not even executed at the first place. So we kept the try and except in that place (inside the while loop)and an else statement counter to the if statement inside the lidar_read function. We got the following outputs. 

Both the instances were observed when we ran the code for a while continuously (No Traceback error this time)
 
1.  " Caught " indicates that lidar_read() was called and is returning the data as it should , for sometime.

![A](https://user-images.githubusercontent.com/81487333/124276951-5011fc80-daf9-11eb-90f5-95af0186de17.png)

2. Here even though we didnt receive a data we set it's value to 0 and printed it. But for the next scan we instead got  "None"  indicating that the lidar_read () was not even called at the first place.

![C](https://user-images.githubusercontent.com/81487333/124276955-51dbc000-daf9-11eb-9d55-648eed73c66f.png)

**Comment by JashwanthKumar252** [comment]
We have successfully identified the method to debug our error keeping in mind the points that you both told about the serial communication. What we are doing now is Closing and opening the serial port thus refreshing it every time we don't get data from the Lidar. When we found this work around, we chose to stress test it to check if the Lidar is stable throughout the test. We ran the code for about 45 mins continuously and we didn't observe any traceback error nor abrupt stopping of the code.

We are able to get an average of 117 readings out of 128 readings that we should get for each scan through 180°. 
The Array  "A" stores our readings and we are printing its length at the end of each scan

![Screenshot from 2021-07-02 09-25-26](https://user-images.githubusercontent.com/81487333/124305672-43ea6700-db1a-11eb-9bab-b288f48ced50.png)
![Screenshot from 2021-07-02 09-26-17](https://user-images.githubusercontent.com/81487333/124305676-451b9400-db1a-11eb-927d-6ace37638b79.png)
![Screenshot from 2021-07-02 09-26-31](https://user-images.githubusercontent.com/81487333/124305682-45b42a80-db1a-11eb-8a59-16e3e5c0cf38.png)
![Screenshot from 2021-07-02 09-27-20](https://user-images.githubusercontent.com/81487333/124305684-464cc100-db1a-11eb-99e0-19698b5aa160.png)

Thanks for the time and your valuable input seniors. @pradhanshrijal @Dcruise546 
We are closing this issue.
